narwhal/primary/src/block_waiter.rs

```diff
@@ -500,339 +86,103 @@ impl<SynchronizerHandler: Handler + Send + Sync + 'static> BlockWaiter<Synchroni
-
-        // Ensure that we mark this block retrieval
-        // as pending so no other can initiate the process
-        self.pending_get_block.insert(id, certificate.clone());
-
-        self.get_block_map_requesters
-            .entry(id)
-            .or_insert_with(Vec::new)
-            .push(sender);
-
-        return Some(fut.boxed());
-    }
-
-    /// This method will ensure that the payload for a block is available to
-    /// the worker before going ahead to retrieve. This is done via the
-    /// block synchronizer handler which will trigger the process of syncing
-    /// if the payload is missing.
-    ///
-    /// # Returns
-    ///
-    /// `true`: If the payload was already synchronized or the synchronization
-    /// was successful
-    /// `false`: When synchronization failed
-    async fn ensure_payload_is_synchronized(&self, certificate: Certificate) -> bool {
-        let sync_result = self
-            .block_synchronizer_handler
-            .synchronize_block_payloads(vec![certificate.clone()])
-            .await;
-
-        sync_result
-            .first()
-            .expect("Expected at least one result back")
-            .is_ok()
-    }
-
-    async fn wait_for_all_blocks(
-        ids: Vec<CertificateDigest>,
-        get_block_receivers: Vec<oneshot::Receiver<BlockResult<GetBlockResponse>>>,
-    ) -> BlocksResult {
-        let result = try_join_all(get_block_receivers).await;
-
-        if let Ok(res) = result {
-            Ok(GetBlocksResponse { blocks: res })
-        } else {
-            Err(BlocksError {
-                ids,
-                error: BlocksErrorType::Error,
+        let block_futures: FuturesOrdered<_> = certificates
+            .into_iter()
+            .map(|(digest, cert)| {
+                self.get_block(digest, cert, successful_payload_sync_set.contains(&digest))
             })
-        }
-    }
+            .collect();
 
-    async fn handle_batch_waiting_result(&mut self, result: BlockResult<GetBlockResponse>) {
-        let block_id = result.as_ref().map_or_else(|e| e.id, |r| r.id);
+        Ok(GetBlocksResponse {
+            blocks: block_futures.collect().await,
+        })
+    }
 
-        match self.get_block_map_requesters.remove(&block_id) {
-            Some(senders) => {
-                for sender in senders {
-                    if sender.send(result.clone()).is_err() {
-                        error!("Couldn't forward results for block {} to sender", block_id)
-                    }
-                }
-            }
-            None => {
-                error!(
-                    "We should expect to find channels to respond for {}",
-                    block_id
-                );
-            }
+    #[instrument(level = "debug", skip_all, err)]
+    async fn get_block(
+        &self,
+        certificate_digest: CertificateDigest,
+        certificate: Option<Certificate>,
+        synced: bool,
+    ) -> BlockResult<GetBlockResponse> {
```

Comment:
The only thing that I have difficulty understanding is how these input params immediately impact if an error is returned vs actually trying to request data from our workers.

Reply:
I'm open to suggestions on refactoring. I put all the error handling in here so that `get_blocks` could just fire everything off and get back a correctly ordered result all in one go by using `try_join_all`. Agreed it is slightly weird to start an operation you know will immediately error out, though it is kind of intuitive if you think of it as "one call to `get_block` == one block response". I added some comments to try to help a little.

Reply:
perfect, thanks for adding some context